### ISSUE TITLE
[git] Update Git Instructions in the docs

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#6837]
On the topic of git-hygiene:  I think we are all in agreement.   Keeping master a clean copy is great for this flow, but accidents do happen, and yes `git pull --rebase` or `git pull --ff-only` are good solutions (and IMO `git config pull.ff only` or `git config pull.rebase only` are even better to save typing).   However my point is that none of this are in the [OpenTitan github guidelines for syncing repos](https://docs.opentitan.org/doc/ug/github_notes#update-your-repository-with-changes-in-the-lowrisc-repo) which advises:
```
$ git checkout master
$ git pull upstream master
$ git push origin
```
Ignoring the disaster that git pull can cause.  I think Mark is set with ff-only as the default. Though I'm just wondering what people's thoughts whether that document could use an update?

_Originally posted by @martin-lueker in https://github.com/lowRISC/opentitan/pull/6749#issuecomment-854937181_

**Comment by eunchan** [comment]
we could add a paragraph or two there for `--ff-only` option to prevent users accidentally create a merged commit. But, general consensus is not to create any local commit on the main branch and always create a new branch for a new PR or any dev works.

So, should we have more detailed github guideline? I am not sure. maybe it is worth to make current doc more clear. However, adding more information may results in indigestibly long github manual. Personally, I think if we think of adding more git related thing, then should consider add clear link to [Pro Git](https://git-scm.com/book/en/v2)

**Comment by rswarbrick** [comment]
Ah, now I understand: thanks, Michael. My preference is to write less, not more. Could we replace that part of the guidelines with a link to some GitHub documentation?

**Comment by martin-lueker** [comment]
I'm in favor of less as well.  Though removing may be overkill.  WD at least really appreciated this documentation. As much as it may seem obvious to you @rswarbrick, there are actually a lot of different ways to organize a repo, and there is some team specific flavor to any chosen workflow. 

For instance even `man git-tutorial` on my Ubuntu system will recommend many things unrelated to our flow (and in fact won't even mention github itself, or the concept of forks, another example of how basic git documentation will be a bit lacking).

For instance the goal of maintaining purely linear commit tree on the master branch, that is by no means universal among Git users, and so IMO we should keep this document live. 

Now as to what I would change: I would just recommend three changes:
1. add `--ff-only` to the line where it says `git pull upstream master`
2. Remind users that the ff-only is helpful for detecting cases when users accidentally add commits to their local master.
3. Point out that `git config pull.ff only` can save users keystrokes on master sync operations.

Two new sentences and an option in the example text.

Does that sound reasonable?

**Comment by rswarbrick** [comment]
My recommendation to anyone I'm teaching to use git is never to use `git pull`. So I'm not really in favour of the changes you suggest. But the current document doesn't really match my recommendations either, so I guess I don't have strong feelings :-)

**Comment by msfschaffner** [comment]
@rswarbrick would you rather add something like
```
git checkout master
git fetch upstream
git reset upstream/master
```
?

**Comment by martin-lueker** [comment]
I personally think `git pull` is fine if you have configured yourself for `ff-only`.
It's enough of a nuisance as it is to tell new people that it takes 3 steps to sync `origin master`,  and this is adding a fourth step for a common operation.  ("Make the common case fast" and all).  In the end though, people should have some discretion about how they manage their private repos as long as the end result is acceptable.   

Furthermore, we won't be able to protect people from *all* git mishaps, so I'd like to discourage making the recommendations more nuanced with the goal of accounting for every possible scenario.  For instance I recently learned that `git push --force` (no other branches listed), forcibly updates everything that git can possibly think of.  I was horrified.  If while we are at it we could eliminate `git push --force ...` that would also be welcome as a precautionary measure, but I'm willing to admit that the tool is imperfect.

If we start discouraging `git pull`, that's fine. Though IMO we can keep it and just add a one-liner for `git config pull.ff only`, and an explanation.  If someone doesn't ever use `pull`, this recommendation will cause said person no harm  ;)

**Comment by rswarbrick** [comment]
> @rswarbrick would you rather add something like
> 
> ```
> git checkout master
> git fetch upstream
> git reset upstream/master
> ```
I'd suggest rebase, not reset: is that what you meant?

But, yeah, I think it's probably better to be a bit explicit. "Power users" who are stuck on the command line can go and learn about `git pull -r` if they want, but I don't think we necessarily have to tell them to use it first.

**Comment by msfschaffner** [comment]
we can make it an explicit `rebase`, yes. I sometimes use `reset` to explicitly align the state of `master` with `upstream/master`, but that (intentionally) gets rid of any commits you might already have on `master`, and could be dangerous if we suggest that here.

so essentially we could replace this set of commands
```
git checkout master
git pull upstream master
git push origin
```
with
```
git checkout master
git fetch upstream
git rebase upstream/master
git push origin
```
to make things explicit, and leave further optimizations in the flow up to "power users".

**Comment by martin-lueker** [comment]
FYI @rasmus-madsen 

Though regarding the rebase, will it appropriately fail if master is corrupted?  The answer is I think: "It depends." 
1. If there is an extra commit on local master, that commit will stay undetected.  
2. However if there is an amended commit, I think this will fail (as it should). 

As much as there is some distaste for the `merge` operations, it is not the merge command, rather but merge *commit* that is the issue we are trying to avoid. 
In this particular instance `git merge --ff-only upstream/master` would be significantly superior in detecting case 1 above.  At which point, why not just let people do a ff-only pull?

**Comment by rswarbrick** [comment]
I think we should *let* people do whatever they fancy (so long as the end results look sane!). If we make any recommendation at all, I think we should *recommend* something that
- won't trash their work (ruling out `git reset`, as Michael said)
- generally makes forward progress (so I prefer `git rebase` over `git merge --ff-only`)

But my preference is still to point at existing GitHub documentation and punt on the issue entirely. Git has been the de-facto VCS in at least the free software world for roughly a decade. It doesn't seem unreasonable to me that we expect contributors to know how to use it (as well as their shell, text editor and the other fundamental bits that you need for programming).

**Comment by rasmus-madsen** [comment]
@martin-lueker
If I understand this correctly the point made is let's avoid regular git pull.
I do agree with this statement. regular git pull can cause all sorts of mess.

if I read the argument correctly: if we should recommend git pull ff-only or git pull -r?

I am not sure I follow your example - so the scenario is that a user accidentally commits work in local master and then pulls?
I don't see anything being undetected!

but in any example all
all `pull --ff-only` will do is nothing!

the next question is then how to recover.
Isn't any solution ending up with a rebase?
in that case, I think we might as well opt for `pull --rebase`

if we are just talking about accidental commits to local master I think -ff-only is good this will force people to move the changes to a branch - reset their local master and then rebase.

but as a general pull recommendation I would go with `pull -r` as it will also be valid for scenarios of a shared branch etc. where do you don't want to move your changes to a new branch

**Comment by msfschaffner** [comment]
This issue has been addressed in #6918.
